controllers/space/space_controller_test.go

```diff
@@ -737,16 +737,126 @@ func TestPromoteSpace(t *testing.T) {
+			Exists().
+			HasTierName(basicTier.Name).
+			Get()
+		require.False(t, tierutil.TierHashMatches(basicTier, nstmplSet.Spec))
+
+		t.Run("requeue while NSTemplateSet is not ready", func(t *testing.T) {
+			// given another round of requeue while NSTemplateSet is *not ready*
+			nsTmplSet.Status.Conditions = []toolchainv1alpha1.Condition{
+				nstemplatetsettest.Updating(), // NSTemplateSet status updated
+			}
+			err := member1.Client.Update(context.TODO(), nsTmplSet)
+			require.NoError(t, err)
+
+			// when
+			res, err := ctrl.Reconcile(context.TODO(), requestFor(s))
+
+			// then
+			require.NoError(t, err)
+			assert.False(t, res.Requeue)
+			spacetest.AssertThatSpace(t, test.HostOperatorNs, "oddity", hostClient).
+				Exists().
+				HasTier(basicTier.Name).
+				HasSpecTargetCluster("member-1").
+				HasStatusTargetCluster("member-1").
+				HasConditions(spacetest.Updating()).
+				HasLabel(tierutil.TemplateTierHashLabelKey(basicTier.Name))
+
+			t.Run("not done when NSTemplateSet is ready within 1s", func(t *testing.T) {
+				// given another round of requeue with NSTemplateSet now *ready*
+				// but LESS than 1s after the Space Ready condition was set to `Ready=false/Updating`
+				nsTmplSet.Status.Conditions = []toolchainv1alpha1.Condition{
+					nstemplatetsettest.Provisioned(),
+				}
+				err := member1.Client.Update(context.TODO(), nsTmplSet)
+				require.NoError(t, err)
+
+				// when
+				res, err := ctrl.Reconcile(context.TODO(), requestFor(s))
+
+				// then
+				require.NoError(t, err)
+				assert.Equal(t, reconcile.Result{
+					Requeue:      true,
+					RequeueAfter: 1 * time.Second}, res) // requeue requested
+				s := spacetest.AssertThatSpace(t, test.HostOperatorNs, "oddity", hostClient).
+					Exists().
+					HasStatusTargetCluster("member-1").
+					HasConditions(spacetest.Updating()).
+					HasLabel(tierutil.TemplateTierHashLabelKey(basicTier.Name)).
+					HasFinalizer().
+					Get()
+
+				t.Run("done when NSTemplateSet is ready for more than 1s", func(t *testing.T) {
+					// given another round of requeue without with NSTemplateSet now *ready*
+					// but MORE than 1s after the Space Ready condition was set to `Ready=false/Updating`
+
+					// hack: change Space's condition timestamp
+					s.Status.Conditions[0].LastTransitionTime = metav1.NewTime(s.Status.Conditions[0].LastTransitionTime.Time.Add(-1 * time.Second))
+					err := hostClient.Status().Update(context.TODO(), s)
+					require.NoError(t, err)
+
+					// when
+					res, err := ctrl.Reconcile(context.TODO(), requestFor(s))
+
+					// then
+					require.NoError(t, err)
+					assert.Equal(t, reconcile.Result{Requeue: false}, res) // no more requeue.
+					spacetest.AssertThatSpace(t, test.HostOperatorNs, "oddity", hostClient).
+						Exists().
+						HasStatusTargetCluster("member-1").
+						HasConditions(spacetest.Ready()).
+						HasLabel(tierutil.TemplateTierHashLabelKey(basicTier.Name)).
+						HasFinalizer()
+				})
+			})
+		})
```

Comment:
I guess that we don't need this sub-tests as all of the cases are already covered in the test-case above (in promotion).
I know that it's not clear for the first glance as it's done as a sub-test of the promotion, but there is nothing new covered here, right?

This is one of the reasons why not nesting such test-cases into sub-tests (and keeping them at the same level) would be much better/clearer. @xcoulon

Reply:
Ok, I think this should do it? e84e7132a0bad6f885cfb3e11cbf3c576716225c